### ISSUE TITLE
Incorrect Pyreverse type annotation

### ISSUE THREAD

**Issue by nickdrozd** [PyCQA/pylint#8266]
### Bug description

`pylint/pyreverse/utils.py` has this function:

```python
def get_annotation_label(ann: nodes.Name | nodes.NodeNG) -> str:
    if isinstance(ann, nodes.Name) and ann.name is not None:
        return ann.name  # type: ignore[no-any-return]
    if isinstance(ann, nodes.NodeNG):
        return ann.as_string()  # type: ignore[no-any-return]
    return ""
```

Argument `ann` is annotated as `nodes.Name | nodes.NodeNG`. According to the types, `ann` is not `None`, so it should be possible to assert this without consequence:

```python
def get_annotation_label(ann: nodes.Name | nodes.NodeNG) -> str:
    assert ann is not None  # <-- assertion added
    if isinstance(ann, nodes.Name) and ann.name is not None:
        return ann.name  # type: ignore[no-any-return]
    if isinstance(ann, nodes.NodeNG):
        return ann.as_string()  # type: ignore[no-any-return]
    return ""
```

But then if Pyreverse is run (say, `pyreverse pylint/pyreverse/`), the assertion fails. Printing `ann` reveals that its value is frequently `None`, contradicting the type annotation.

I don't know if this is a Mypy bug, or a configuration issue, or what. I also don't know if the problem is limited to this function or if it goes further.

**Comment by Pierre-Sassoulas** [comment]
mypy trust us if we tell it that there's no None. Some function in astroid says None is never returned but they're lying and then we crash (Recent example: https://github.com/PyCQA/pylint/issues/8109). As you know mypy can't even be enforced in astroid yet. So on the pylint side what we can do is change the typing and handle all cases accordingly.

**Comment by DanielNoord** [comment]
To expand on @Pierre-Sassoulas, everything that involves `astroid` imports is `Any` as far as `mypy` is concerned. This is because `astroid` doesn't have a `py.typed`.

**Comment by nickdrozd** [comment]
I see, so as far as Mypy is concerned, the nodes in the signature could anything, including `None`? Sounds like Mypyc compilation is further off than I realized :thinking: 

Anyway, the problem is well understood, so I'll close this.
